### ISSUE TITLE
Setting "Allow Google login for extensions" does not work

### ISSUE THREAD

**Issue by mariospr** [brave/brave-browser#15754]
## Test plan
See https://github.com/brave/brave-core/pull/14069

## Description 

Today, @AlexeyBarabash  and me have found what seems to be a bug in the "Allow Google logins for extensions" feature, or at least in how it should work according to the [information published in the wiki](https://github.com/brave/brave-browser/wiki/Allow-Google-login---Third-Parties-and-Extensions#extensions): _"When this option is enabled: It enables chrome.identity for extensions so extensions like Google Keep and Google Calendar can retrieve an OAuth token from google to authenticate users."_ 

We know that this worked in Brave 1.14  (~9 months ago) because @jumde landed [PR #5616](https://github.com/brave/brave-core/pull/5616) precisely to enable extensions to use Google's login...  and even on Brave 1.20 (~5 months ago) because @AlexeyBarabash landed [PR #7346](https://github.com/brave/brave-core/pull/7480) at that time and keeping extensions logging in with Google auth was one of the things he kept in mind while doing that.

Just to make sure there was not a problem with the extensions themselves we tried this with Google Chrome 90.0.4430.72 and it worked perfectly: 

1. Install Google Keep extension
2. Enable "Allow Google Login for extensions" and relaunch
3. Navigate to a random page: I loaded https://arstechnica.com/
4. Log to your Google account (either from a google.com site of clicking on the extension and then clicking on "Sign in" in the popup")
5. Go back to the random page loaded in (3) and click on the extension's icon
6. After the previous step, it is possible enter a note in Google Keep right from the extension's popup

## Steps to Reproduce

1. Install Google Keep extension
2. Enable "Allow Google Login for extensions" and relaunch
3. Navigate to a random page: I loaded https://arstechnica.com/
4. Log to your Google account (either from a google.com site of clicking on the extension and then clicking on "Sign in" in the popup")
5. Go back to the random page loaded in (3) and click on the extension's icon

OTHER INFO:
* Steps 1-5 fail on Brave regardless of whether you have Brave Sync enabled or not
* Steps 1-5 from Google Chrome works as expected, even if you don't turn Google Sync on

## Actual result:

The extension's popup keeps asking you to "Sign in", this is how it looks on Brave 1.23.71:
![brave](https://user-images.githubusercontent.com/2197259/117844315-e0f9f380-b27f-11eb-9a4c-93a5766627c2.jpg)

## Expected result:

You should be able enter a note in Google Keep right from the extension's popup, as it happens on Chrome 90.0.4430.72:
![chrome](https://user-images.githubusercontent.com/2197259/117844391-ee16e280-b27f-11eb-9328-11074ccad33e.jpg)

## Reproduces how often: 

Always

## Desktop Brave version:

Brave 1.23.71

## Version/Channel Information:
- Can you reproduce this issue with the current release? Yes
- Can you reproduce this issue with the beta channel? Yes
- Can you reproduce this issue with the nightly channel? Yes

## Other Additional Information:

- Does the issue resolve itself when disabling Brave Shields? No
- Does the issue resolve itself when disabling Brave Rewards? No
- Is the issue reproducible on the latest version of Chrome? No

## Miscellaneous Information:

Related past issues:
* [Unable to authenticate or use Google Keep extension in Brave #3650](https://github.com/brave/brave-browser/issues/3650)
* [Google Calendar Extension won't authenticate #4672](https://github.com/brave/brave-browser/issues/4672)

**Comment by mariospr** [comment]
Some extra info after debugging with @AlexeyBarabash for a bit:

It looks like this could be caused because of `profile->GetPrefs()->GetBoolean(prefs::kSigninAllowed)` always returning `false` at `src/chrome/browser/extensions/api/identity/identity_get_auth_token_function.cc`:
```c++
bool IsBrowserSigninAllowed(Profile* profile) {
  return profile->GetPrefs()->GetBoolean(prefs::kSigninAllowed);
}
```

Another interesting data point is that 7 months ago the following landed in ToT:
https://chromium-review.googlesource.com/c/chromium/src/+/2429063

...which could have an impact since now "is signing allowed" depends on the value of `google_apis::HasOAuthClientConfigured()`, which will return `true` or `false` depending on the API keys. Still not 100% sure whether that's the reason of this, though, but the timeline kind of makes sense as being landed 7 months ago could explain why we didn't detect it when @AlexeyBarabash landed [PR #7346](https://github.com/brave/brave-core/pull/7480) back in December 2020.

**Comment by mariospr** [comment]
@jumde shared why this is happening now, and the plan in place to get it fixed:

The problem was caused by recent changes to the policy for usage of OAuth credentials: currently, Google sign in for extensions is capped to a few thousand users for Chromium derivatives, which is why we're getting the error even if we have valid OAuth credentials.

Fortunately, there's already [work happening upstream to allow derivatives to use the chrome.identity API for extensions](https://chromium-review.googlesource.com/c/chromium/src/+/2704571) which will allow to get this problem fixed in Brave once that change lands upstream.

Therefore, this issue is waiting for that change to land before it's possible to get it fixed. 

Last, whenever this gets fixed please make sure it does not introduce a second instance of IdentityManagerFactory, which was causes issues from Cr92 forward, see issue https://github.com/brave/brave-browser/issues/15659 (for which we've already got a PR that will be landing soon, hopefully today: https://github.com/brave/brave-core/pull/8731)

**Comment by santiyounger** [comment]
Thanks for the info in here, good to know I'm not the only one with this issue. After already enabling `Allow Google login for extensions` and updating Brave, I keep getting this message
![image](https://user-images.githubusercontent.com/51902730/121811751-f1104480-cc2a-11eb-8325-21e588fbc609.png)

I click on the Sign in button. It takes me to a new Google login page. I'm able to sign in, but after refreshing the page  I want to save (or opening the new one. 

it shows the same message again 

![image](https://user-images.githubusercontent.com/51902730/121811790-0e451300-cc2b-11eb-8cde-bacf1d9c242b.png)

I'm on Garuda Linux (Arch Based)

Thanks

**Comment by qdgiang** [comment]
I'm also in the same spot as @santiyounger. I have already tried user-agent switch extension and enabled all the mentioned settings, and it keeps asking me to log in again.

I'm on Windows 10

**Comment by santiyounger** [comment]
I saw there are a lot of reported issues on this topic, not sure which one is the best to follow updates on. 

If anyone has a link of the main issue where this is being tracked I'd love to know. 

Sadly because I've started to make heavy use of Google Keep. I had to sell my soul and move to Google Chrome until this issue can be fully solved in Brave. 

I'm sorry for being a traitor :rofl:

**Comment by SilverWx** [comment]
As a user I can say I have the problem too

**Comment by santiyounger** [comment]
> As a user I can say I have the problem too

It's a shame a simple issue like this made Brave no longer an option for me. I'll keep an eye on this issue to see if I can return to Brave one day.

**Comment by mariospr** [comment]
@jumde @iefremov Pinging you on this issue to make sure you're aware of these latest reports.

**Comment by iefremov** [comment]
Thanks @mariospr , I wasn't aware of this. Probably we could bump the priority @rebron ?

**Comment by Michael-F-Ellis** [comment]
I have the problem, too.  
- Google Keep extension version 4.21292.540.1 
- Brave Version 1.27.108 Chromium: 92.0.4515.107 (Official Build) (x86_64).

**Comment by Tonev** [comment]
Another user having the same issue as @mariospr (actual & expected result screenshots).

https://community.brave.com/t/cant-sign-in-to-google-keep-extension/271598/22

**Comment by davenir** [comment]
I have this issue too:

Brave version :  1.27.111 Chromium: 92.0.4515.131 (Official Build) (64-bit)
Keep version:    4.21302.540.1
Windows 10

**Comment by fabricapo** [comment]
I have this issue too (hanging on to Brave hoping you'll fix this soon...)
Brave: V1.28.105 on Mac OS 11.2.3

**Comment by jumde** [comment]
No updates from our friends at chromium on this change - https://chromium-review.googlesource.com/c/chromium/src/+/2704571 - Requested an updated, will share details here when available. Apologies for the delay here.

**Comment by jwashburn** [comment]
Having the same issue. Spent the last couple hours reading up on it. Following this thread now. Thank you to everyone who's helping out on this!

**Comment by creep33** [comment]
> Having the same issue. Spent the last couple hours reading up on it. Following this thread now. Thank you to everyone who's helping out on this!

Have you fixed it?

**Comment by jwashburn** [comment]
> > Having the same issue. Spent the last couple hours reading up on it. Following this thread now. Thank you to everyone who's helping out on this!
> 
> Have you fixed it?

No. Still not working for me. On the Google keep extension in Brave, it prompts me to sign in, but when I do, nothing happens.

**Comment by hiraksarkar** [comment]
Not working for me either.

**Comment by HouserR** [comment]
Same.
Brave V1.29.77
Linux Mint 20.2

**Comment by saadatfar** [comment]
Same.
Brave V1.29.80
Windows 10

**Comment by bsclifton** [comment]
Google Keep (and some other extensions) use a 1st party Google API that will not be available to Chromium forks (Brave, Edge, Vivaldi, etc).

Details were shared back in January 2021 via blog (https://blog.chromium.org/2021/01/limiting-private-api-availability-in.html). The `embedder-dev` mailing list had discussions too - and in February, Google shared a patch which allows chrome.identity api to be used for 3rd party APIs:
https://chromium-review.googlesource.com/c/chromium/src/+/2704571

Developers at Google confirmed tha the old 1st party auth (previously available until it was shut down March 21) is considered `wontfix` by Google ☹️ 

We are working on updating the the code to allow `chrome.identity` to work with Google APIs meant for 3rd party (see patch by @mariospr in https://github.com/brave/brave-core/pull/10103).  This will allow extensions which use Google Logins to work - but unfortunately won't solve cases where 1st party authentication is needed

**Comment by mwmentor** [comment]
Hi there:  I am running Brave [Version 1.36.112 Chromium: 99.0.4844.51 (Official Build) (x86_64)](https://brave.com/latest/) and this is still a problem with Google Keep... Thanks.

**Comment by mariospr** [comment]
CC @jumde @bsclifton

**Comment by legolasdimir** [comment]
> Google Keep (and some other extensions) use a 1st party Google API that will not be available to Chromium forks (Brave, Edge, Vivaldi, etc).
> 
> Details were shared back in January 2021 via blog (https://blog.chromium.org/2021/01/limiting-private-api-availability-in.html). The `embedder-dev` mailing list had discussions too - and in February, Google shared a patch which allows chrome.identity api to be used for 3rd party APIs: https://chromium-review.googlesource.com/c/chromium/src/+/2704571
> 
> Developers at Google confirmed that non-Google access to 1st party APIs using `chrome.identity` (previously available until it was shut down March 21) is considered `wontfix` by Google ☹️
> 
> We are working on updating the the code to allow `chrome.identity` to work with Google APIs meant for 3rd party (see patch by @mariospr in [brave/brave-core#10103](https://github.com/brave/brave-core/pull/10103)). This will allow extensions which use Google Logins to work - but unfortunately won't solve cases where 1st party authentication is needed

If you look at the post I quoted above, I believe that sadly, Brave is hosed with Google keep. It's not working in ANY chromium browser except edge now because of changes Google made to their API.

**Comment by stephendonner** [comment]
## Verification `PASSED` using

### Steps:

1. installed `1.43.56`
2. launched Brave
3. loaded `https://chrome.google.com/webstore/detail/drive-sample/jpabeekbjicamajjcfejnochhmlbpgjh/related?pli=1`
4. clicked on `Add to Brave`
5. clicked on `Add app`
6. double-clicked on `Drive Sample`
7. clicked on `Authorize`
8. entered my Google credentials

#### Confirmed I was able to upload files to my Google Drive by dragging and dropping them on the `Drive Sample` app.

example | example | example | example | example | example | example
---------|----------|---------|-----------|---------|---------|---------
<img width="1312" alt="Screen Shot 2022-08-03 at 11 26 46 AM" src="https://user-images.githubusercontent.com/387249/182683814-c34d6cac-b4ba-45d5-93dd-8db9264519a0.png"> | <img width="1312" alt="Screen Shot 2022-08-03 at 11 27 14 AM" src="https://user-images.githubusercontent.com/387249/182683831-008e5f51-c4c2-4481-9f36-816ecf708ff7.png"> | <img width="1312" alt="Screen Shot 2022-08-03 at 11 27 04 AM" src="https://user-images.githubusercontent.com/387249/182683826-c40db4f1-3fb9-4193-a971-fb50c84d33b4.png"> | <img width="612" alt="Screen Shot 2022-08-03 at 11 27 21 AM" src="https://user-images.githubusercontent.com/387249/182683838-7fc20c9c-144a-4b32-87b8-0c6b5071c3e7.png"> |  <img width="1136" alt="Screen Shot 2022-08-03 at 11 28 16 AM" src="https://user-images.githubusercontent.com/387249/182683842-7bfa7af9-b75d-4b0c-9d65-8051ee6169e4.png"> | <img width="612" alt="Screen Shot 2022-08-03 at 11 28 29 AM" src="https://user-images.githubusercontent.com/387249/182683844-a44e5c81-6f03-40b3-9e6b-a10eeffcdd29.png"> | <img width="1312" alt="Screen Shot 2022-08-03 at 11 42 02 AM" src="https://user-images.githubusercontent.com/387249/182685250-9d21cd8f-9346-4c5e-b0a5-7b03ec9a2a12.png"> | <img width="1312" alt="Screen Shot 2022-08-03 at 11 42 02 AM" src="https://user-images.githubusercontent.com/387249/182685803-2445328c-cdb1-4374-aab7-cd7f4f597d87.png">

**Comment by legolasdimir** [comment]
I tried the google keep extension on the nightly version (1.44.37) on mac os Montery 12.5 and i get a  pop up when i click sign in as well as when the extension loads that throws an error  stating invalid uri scheme. If i click go to keep in the extension menu it asks you to sign in which i can do and it then takes me to keep but if i go back to a page and click the keep icon i still get the 'you must sign in message'

**Comment by TrySpace** [comment]
Confirmed on nightly 1.44.40 on Windows 10 (Version	10.0.18363 Build 18363) I can now login to a google account with extension. 
Previously I did not get a google login popup, now I do (although the popup is as large as the browser window)

**Comment by legolasdimir** [comment]
so essentially what i feared. keep is broken on anything that's not chrome

**Comment by UjCbFwtBayFM** [comment]
> ### Also tested with Google Keep:
> example	example	example	example
> <img alt="Screen Shot 2022-08-03 at 11 54 40 AM" width="1312" src="https://user-images.githubusercontent.com/387249/182687921-bf2d5aec-1b0a-41a7-8ddc-a540f2a0785e.png">	<img alt="Screen Shot 2022-08-03 at 11 55 12 AM" width="1312" src="https://user-images.githubusercontent.com/387249/182687942-03f23787-4835-454c-aa1d-66cbb6e59294.png">	<img alt="Screen Shot 2022-08-03 at 11 55 19 AM" width="1418" src="https://user-images.githubusercontent.com/387249/182687949-c1cae5b3-a3e9-4a7b-a59a-683803ba9804.png">	<img alt="Screen Shot 2022-08-03 at 11 55 22 AM" width="1312" src="https://user-images.githubusercontent.com/387249/182687956-4e5cc9fc-38dd-4e33-913c-6a761ebd8fa6.png">

What did you test exactly there?

**Comment by spylogsster** [comment]
@UjCbFwtBayFM PTAL https://github.com/brave/brave-browser/issues/15754#issuecomment-920514585

**Comment by erick-jeronimo** [comment]
This issue was solved? I'm having the same problem

**Comment by legolasdimir** [comment]
@erick-jeronimo are you talking about keep not letting you save? The recently referenced comment basically says they can't fix it. Google changed their API so extensions that require a certain form of login won't work outside of chrome anymore. So we're just hosed.i went back to chrome because of this. And reliability issues with sending pages between devices

**Comment by erick-jeronimo** [comment]
yes @legolasdimir, thanks a lot for the clarification!

**Comment by kdenhartog** [comment]
> I have noticed an annoyance/deviation with this from google-chrome: I have to re-enter my google credentials, including the username, whereas normally (in chrome) I get the dialog to select the account, and don't have to enter my password.

This is to be expected from my understanding of how we decided to implement this versus Google. In Google's case they grant themselves an exception because Chrome is given access to restricted APIs which make it possible for their browser to effectively behave like the client in OAuth rather than treating each individual extension in that way.

You can read more into this from their [blog post](https://blog.chromium.org/2021/01/limiting-private-api-availability-in.html) and the [chromium docs](https://www.chromium.org/developers/how-tos/api-keys/#polyfilling-chromeidentity-api-in-your-chromium-derivative).

**Comment by hardwood6341** [comment]
Same Problem
